### ISSUE TITLE
optimizations

### ISSUE THREAD

**Issue by Estevo-Aleixo** [qarl/Painter#1]
How would I change the image comparison to only look at the bounding box of the change in the generated image and the mona image? Something like UIRectClip ( CGRect rect ); ?

Instead of choosing the color before comparison why not make the comparison/iteration return the average color of the bounding box? (or of the center vertical and horizontal line) Including another comparison to the image would make it a 2n problem (without the optimization of the center vert/horiz line), but would this comparison be useful?

**Comment by qarl** [comment]
i think both of those ideas would significantly speed up the convergence.  as it is, the current code is not optimized for speed - it's optimized for simplicity.

in the first case, you'd have to be careful: the score for one bounding box would be different than for another bounding box.  so you couldn't just keep last iteration's score and compare against that - you'd need to recompute the previous image's score for each new bounding box.

**Comment by Estevo-Aleixo** [comment]
Would making the generated image the same size as the mona image remove the need for every compare to resize both images? (I am not an objective-c programmer, and I do not know the API well enough yet) (or whichever is bigger between the mona image and the canvas)

This may be a bug, but if the pixel is three color channels mixed into one representation of a pixel is it correct to do a abs(505152 - 006000) ? where the red channel is 0x50 green is 0x51 blue is 0x52. Would an XOR be a better representation of the difference in channels?

**Comment by qarl** [comment]
yup.  resizing the Mona image for every comparison is very wasteful, time-wise.   :)

**Comment by Estevo-Aleixo** [comment]
The decision and change idea (without the xor for separating color channels)

```
cur_score = mona_score - (mona_aabb_score - change_aabb_score)
if cur_score > mona_score:
    paint()
    mona_score = cur_score
```
